lib: smf: Add State Machine Framework

sambhurst opened the pull request:
Add an application agnostic State Machine Framework library to
Zephyr that provides an easy way for developers to integrate
state machines into their application.

Twister passed:
twister -T tests/lib/smf/

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
lib: smf: Add State Machine Framework

Add an application agnostic State Machine Framework library to
Zephyr that provides an easy way for developers to integrate
state machines into their application.

Twister passed:
twister -T tests/lib/smf/

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst commented on tests/lib/smf/src/test_lib_smf.c:
```diff
@@ -0,0 +1,350 @@
+ *
+ *
+ *
+ * Hierarchical Test Transition:
```
Done

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+#define _SMF_H__
+
+/**
+ * @brief Macro to create a hierarchical state that prints a message
```
Done

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+ * @param x  State exit function
+ * @param p  State parent object or NULL
+ */
+#define HSM_CREATE_MSG_STATE(me, mx, e, r, x, p) \
```
Done

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+#define HSM_CREATE_MSG_STATE(me, mx, e, r, x, p) \
+{ \
+	.entry_msg = me, \
+	.exit_msg  = mx, \
```
I'm not sure how to do this. Place the guard inside a define

mbolivar-nordic replied:
> I'm not sure how to do this. Place the guard inside a define

You can use IF_ENABLED for this.

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+typedef void (*state_execution)(void *obj);
+
+/*
+ * General state that can be used in multiple state machines.
```
Done

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+};
+
+/**
+ * Changes a state machines state. This handles exiting the previous state and
```
Done

sambhurst commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,150 @@
+void set_state(struct smf_ctx *ctx, const struct smf_state *new_state);
+
+/**
+ * Runs one iteration of a state machine (including any parent states)
```
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+This example turns the following state diagram into code using the SMF, where
+the initial state is S0.
+
+.. figure:: img/flat.png
```
note: you may want to use `graphviz` diagrams, they are natively supported in the documentation (render to SVG and are better viewed in both light/dark modes)

sambhurst replied:
Yes, the image is much better. Thanks.

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+doesn't make sense and will generate a warning.
+
+State Machine Execution
+=======================
+To run the state machine, the **run_state** function should be called in some
```
```suggestion
=======================

To run the state machine, the **run_state** function should be called in some
```

sambhurst replied:
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+**void run_state(smf_ctx *ctx)**
+
+Flat State Machine Example
+==========================
+This example turns the following state diagram into code using the SMF, where
```
```suggestion
==========================

This example turns the following state diagram into code using the SMF, where
```

sambhurst replied:
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+		}
+	}
+
+When designing hierarchical state machines, the following should be considered\:
```
```suggestion
When designing hierarchical state machines, the following should be considered:

```

sambhurst replied:
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+	}
+
+When designing hierarchical state machines, the following should be considered\:
+    - Ancestor entry actions are executed before the sibling entry actions. For
```
```suggestion
- Ancestor entry actions are executed before the sibling entry actions. For
```

sambhurst replied:
Done

gmarull commented on lib/smf/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+zephyr_library()
+
+zephyr_include_directories(.)
```
public API should likely be placed in `include/`, so this would not be required.

sambhurst replied:
Done

gmarull commented on lib/smf/Kconfig:
```diff
@@ -0,0 +1,24 @@
+	  Enabling this option will not allow any parent states, resulting
+	  in a Flat State Machine.
+
+config SMF_PRINT_MSG
```
isn't logging sufficient? Should that option depend on logging being enabled?

sambhurst replied:
Yes I agree that the message printing is unnecessary because developers can use logging.

AndreyDodonov-EH replied:
@sambhurst I see that this option is not checked anywhere, is giving output about states to-be-implemented?
I smf.c the only LOG usage is 
`LOG_WRN("Calling %s from exit action", __func__);`

gmarull commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,217 @@
+ */
+struct internal_ctx {
+	const struct smf_state *last_entered;
+	uint32_t running : 1;
+	uint32_t enter	 : 1;
+	uint32_t exit	 : 1;
```
since they seem to be boolean flags, `bool running: 1` could be used

sambhurst replied:
Done

gmarull commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,217 @@
+	const struct smf_state *orig_b = b;
+
+	/* There are no common ancestors */
+	if (b == NULL)
+		return NULL;
```
```suggestion
	if (b == NULL) {
		return NULL;
	}
```
missing braces

sambhurst replied:
Done

gmarull commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,217 @@
+	/* This assumes that both A and B are NULL terminated without cycles */
+	while (a != NULL) {
+		/* We found a match return */
+		if (a == b)
```
ditto

sambhurst replied:
Done

gmarull commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,217 @@
+	if (current->entry) {
+#ifdef CONFIG_SMF_PRINT_MSG
+		if (current->entry_msg) {
+			printk("%s\n", current->entry_msg);
```
it's better to use logging subsystem for that.

sambhurst replied:
This has been removed in favor of the logging system

gmarull commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,217 @@
+							const struct internal_ctx *const internal,
+							const struct smf_state *current)
+{
+	if (!current)
+		return;
+
+	/* If set_state is called during run, don't call remain functions. */
+	if (!internal->running)
+		return;
+
+	if (current->run)
+		current->run((void *)ctx);
```
missing braces

sambhurst replied:
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+	static void parent_exit(void *o) {}
+
+	/* State S0 */
+	static void s0_entry(void *o) {}
```
would it be possible to allow `NULL` entry/exit callbacks if they are a no-op?

sambhurst replied:
I made some NULL and added comments to make things clear.

gmarull commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,160 @@
+/*
+ * General state that can be used in multiple state machines.
+ *
+ * entry - Optional method that will be run when this state is entered
+ * run   - Optional method that will be run repeatedly during state machine
+ *			loop
+ * exit  - Optional method that will be run when this state exists
+ * parent- Optional parent state that contains common entry/run/exit
+ *	implementation among various child states.
+ *	entry: Parent function executes BEFORE child function.
+ *	run:   Parent function executes AFTER child function.
+ *	exit:  Parent function executes AFTER child function.
+ *
+ *	Note: When transitioning between two child states with a shared parent,
+ *	that parent's exit and entry functions do not execute.
+ */
```
These docstrings are not Doxygen, each field should be documented individually,

```c
/**What is struct x */
struct x {
    /** What is y */
    int y;
}
```

sambhurst replied:
Done

gmarull commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,160 @@
+#endif
+};
+
+/* Defines the current context of the state machine. */
+struct smf_ctx {
+	const struct smf_state *current;
+	const struct smf_state *previous;
+	intptr_t internal[2];
+};
```
missing docs

sambhurst replied:
Done

gmarull commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+
+		/* Run the state machine */
+		while(1) {
+			run_state(SMF_CTX(s_obj));
```
should error handling be allowed?

sambhurst replied:
I added a way to return an error and terminate the state machine

gmarull reviewed (commented):
I like the idea, some initial comments/thoughts

keith-zephyr commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,160 @@
+	.parent = _parent \
+}
+
+#else
```
Is it a supported use case to have CONFIG_SMF_FLAT=n, but enable a flat state machine.  In other words, is it supported to enable both hierarchical and flat state machines in the same build?

sambhurst replied:
The short answer is no. Initially I was going to just have the hierarchical state and a designer could leave all the parent entries NULL if they wanted a purely flat state machine but I thought that would waste memory, so that's why I added the CONFIG_SMF_FLAT macro.

keith-zephyr commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+**HSM_CREATE_MSG_STATE(me, mx, e, r, x, p)** is used to create a hierarchical
+state that can have an optional parent state and prints a message in the
+optional Entry action and a message in the optional Exit action. The macro has
+the following parameters:
```
Note that the parameters are now out of date.  Probably better to document the parameters in the code only and just provide a summary of the macro and it's intended use here.

mbolivar-nordic replied:
I would suggest looking into the `:c:macro:` role and using it as needed to link to the API docs for this instead.

sambhurst replied:
Thanks for the suggestion

keith-zephyr commented on lib/smf/smf.h:
```diff
@@ -0,0 +1,160 @@
+	.entry     = _entry,     \
+	.run       = _run,       \
+	.exit      = _exit       \
+}
```
I think you'll need to wrap this macro with CONFIG_SMF_PRINT_MSG so that users can cleanly disable the entry/exit messages with just a config change:

```
#ifdef CONFIG_SMF_PRINT_MSG
#define FSM_CREATE_MSG_STATE() \
   { .entry_msg = _entry_msg, .exit_msg = _exit_msg, .entry = _entry, /* etc */ }
#else
#define FSM_CREATE_MSG_STATE() \
   { .entry = _entry, /* etc */ }
#endif
```

sambhurst replied:
I've removed all the code related to PRINT_MSG in favor of using the logging subsystem

mbolivar-nordic commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+
+The State Machine Framework (SMF) is an application agnostic framework that
+provides an easy way for developers to integrate state machines into their
+application. The framework can be added to any project by including the
+CONFIG_SMF=y into the projects prj.conf file.
```
```suggestion
application. The framework can be added to any project by enabling the
:kconfig:`CONFIG_SMF` option.
```

sambhurst replied:
Done

mbolivar-nordic commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,267 @@
+A state is represented by three functions, where one function implements the
+Entry actions, another function implements the Run actions, and the last
+function implements the Exit actions. The prototype for these functions is as
+follows: **void funct(void *obj)**, where the **obj** parameter is a user
```
Please read through the documentation guidelines and make adjustments: https://docs.zephyrproject.org/latest/contribute/documentation/index.html

E.g. this should be using double backticks for inline code literals as follows:

```
follows: ``void funct(void *obj)`` [...]
```

sambhurst replied:
Done

AndreyDodonov-EH commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,269 @@
+	}
+	static void s0_run(void *o)
+	{
+		smf_set_state(SMF_CTX(s_obj), &demo_states[S1]);
```
@sambhurst I think & is needed, same as in other SMF_CTX calls

sambhurst replied:
Done

AndreyDodonov-EH commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,269 @@
+	}
+
+	/* Populate state table */
+	static const struct smf_state demo_states[] {
```
"`=`" is missing:
`static const struct smf_state demo_states[] = `

sambhurst replied:
Done

AndreyDodonov-EH commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,269 @@
+	}
+
+	/* Populate state table */
+	static const struct smf_state demo_states[] {
```
I guess also = missing

sambhurst replied:
Done

sambhurst replied:
Done

AndreyDodonov-EH commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,269 @@
+	/* Populate state table */
+	static const struct smf_state demo_states[] {
+		/* Parent state does not have a run action */
+		[PARENT] = HSM_CREATE_STATE(parent_entry, NULL, parent_exit, NULL);
```
Should be , not ;

sambhurst replied:
Done

AndreyDodonov-EH commented on doc/guides/smf/index.rst:
```diff
@@ -0,0 +1,269 @@
+		/* Run the state machine */
+		while(1) {
+			/* State machine terminates if a non-zero value is returned */
+			ret = smf_run_state(SMF_CTX(s_obj);
```
missing )

sambhurst replied:
Done

AndreyDodonov-EH reviewed (changes requested):
@sambhurst
1. Could you please make sure that examples in index.rst are compilable?

2. Could you add logging on state changes? Haven't found any, or am I missing something?
3. Is there a visualization tool to generate from this state machine? If yes, would be nice to document it.
4. Considering Zephyrs threading features, wouldn't it be more useful to have event-driven state-machine? 
    I.e. defining which events trigger which transitions....

sambhurst commented:
@AndreyDodonov-EH

**1. Could you please make sure that examples in index.rst are compilable?**
I verified that both examples compile and run as expected.

**2. Could you add logging on state changes? Haven't found any, or am I missing something?**
I didn't think it was worth having the additional macros that print messages with the logging systems because I figured the developer could just add there own log prints where they see fit. My earlier replies were not clear on my intentions.

**3) Is there a visualization tool to generate from this state machine? If yes, would be nice to document it.**
I used graphviz dot.
dot -Tsvg flat.gv -o flat.svg
I'll add to dot scripts to the guide.

**4) Considering Zephyrs threading features, wouldn't it be more useful to have event-driven state-machine?**
I was thinking the developer could use the event system if they want to. For example, something like the following:

while(1) {
   k_poll(events, 1, K_FOREVER);
   /* set the event in the user object and pass to the state machine */
   user_obj->events = events;
   smf_run_state(SMF_CTX(user_obj));
}

AndreyDodonov-EH commented:
> @AndreyDodonov-EH
> 
> **1. Could you please make sure that examples in index.rst are compilable?**
> I verified that both examples compile and run as expected.
> 
> **2. Could you add logging on state changes? Haven't found any, or am I missing something?**
> I didn't think it was worth having the additional macros that print messages with the logging systems because I figured the developer could just add there own log prints where they see fit. My earlier replies were not clear on my intentions.
> 
> **3) Is there a visualization tool to generate from this state machine? If yes, would be nice to document it.**
> I used graphviz dot.
> dot -Tsvg flat.gv -o flat.svg
> I'll add to dot scripts to the guide.
> 
> **4) Considering Zephyrs threading features, wouldn't it be more useful to have event-driven state-machine?**
> I was thinking the developer could use the event system if they want to. For example, something like the following:
> 
> while(1) {
> k_poll(events, 1, K_FOREVER);
> /* set the event in the user object and pass to the state machine */
> user_obj->events = events;
> smf_run_state(SMF_CTX(user_obj));
> }

@sambhurst 
1. First things first, thanks a lot for the fixes! :-) 
2. I agree, logging can be added by developers as necessary
3. Great, thanks a lot!
4. Currently, we have to set states in sX_run bodies, so it becomes somewhat hard to have an overview of the logic.
    And if we add events to it, it becomes even harder to grasp the transitions.
    What I meant is that there is no way to define transitions table-like.

sambhurst commented:
@AndreyDodonov-EH

I initially considered a state table approach but thought  it only useful for simple state machines because of the upfront effort needed to setup the table of events / states and making changes to the table could be error prone. Also, the tables are usually sparse and this would waste memory.

I think the developer will need to have good documentation that describes the state transition that occur in the entry and run actions.

AndreyDodonov-EH commented:
@sambhurst 
I agree that making changes to the table can be error-prone, but it can be covered with compile-time and initializing phase checks.
I.e. when initializing state machine, table's validity can and has to be proven.

Unfortunately, documentation can quickly become obsolete, especially if transitions are not in one place, but scattered across functions, which can be edited frivolously.
In contrast, having table in the code provides overview right away.

I agree that it increases memory footprint, but function pointers and event ids do not occupy that much space (okay, depends on the application, but Zephyr anyway is not meant for ultra-low memory applications, AFAIK)

alexanderwachter commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,246 @@
+	}
+
+#ifndef CONFIG_SMF_FLAT
+	call_entry_functions(ctx, internal, stop, current->parent);
```
Please no recursion.
The problem is that recursion can blow your stack easily.
It is also not obvious that for a flat state machine, that the recursion stops immediately because stop is NULL

sambhurst replied:
I've removed the recursion

alexanderwachter commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,246 @@
+	}
+
+#ifndef CONFIG_SMF_FLAT
+	call_exit_functions(ctx, internal, stop, current->parent);
```
Same applies here

sambhurst replied:
removed

alexanderwachter commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,246 @@
+	}
+
+#ifndef CONFIG_SMF_FLAT
+	call_run_functions(ctx, internal, current->parent);
```
And here

sambhurst replied:
removed

keith-zephyr commented on include/smf.h:
```diff
@@ -0,0 +1,156 @@
+	 * The state machine casts this to a "struct internal_ctx" and it's
+	 * used to track state machine context
+	 */
+	intptr_t internal[2];
```
Why is this an array?  This only gets cast to a 32-bit bitmask.

sambhurst replied:
This is very misleading so I've changed it to a uint32_t without the array.

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,308 @@
+	ctx->previous = NULL;
+	ctx->terminate_val = 0;
+
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
```
I suggest breaking this block out as a separate function. Something like ``static void smf_set_initial_ancestor(struct smf_ctx *ctx, const struct smf_state *init_state);``

If you do this and add the helper Kconfig symbol, then you can avoid the #ifdefs completely.

sambhurst replied:
Done

keith-zephyr commented on lib/smf/Kconfig:
```diff
@@ -0,0 +1,28 @@
+
+if SMF
+
+config SMF_ANCESTOR_NUM
```
I suggest you create a helper symbol:

```
config SMF_ANCESTOR_SUPPORT
    def_bool SMF_ANCESTOR_NUM > 0
    help
       if y, then the state machine framework includes ancestor state support
```

This allows you to replace the references of ``#if (CONFIG_SMF_ANCESTOR_NUM > 0)`` with ``#ifdef CONFIG_SMF_ANCESTOR_SUPPORT`` and ``IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT``.

sambhurst replied:
done

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,308 @@
+		return ctx->terminate_val;
+	}
+
+#if (CONFIG_SMF_ANCESTOR_NUM > 0)
```
Move this block into a separate function.

sambhurst replied:
done

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,319 @@
+struct internal_ctx {
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
+	bool new_state : 1;
+#endif
```
If you allow the new_state flag unconditionally, I believe you can remove the #ifdef around the ``smf_execute_ancestor*`` routines.  You can then also convert all the calls to the ``smf_execute_ancestor*`` routines using IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT).

You'll have to add the __unused attribute to these routines to avoid a warning when CONFIG_SMF_ANCESTOR_SUPPORT=n

sambhurst replied:
Done

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,319 @@
+	if (smf_execute_ancestor_entry_actions(ctx, init_state)) {
+		return;
+	}
+#endif
```
See previous comment,  if new_state is allowed when CONFIG_SMF_ANCESTOR_SUPPORT=n, then this can use IS_ENABLED(), which is the preferred style when possible.

```
if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
    internal->new_state = false;
    if (smf_execute_ancestor...)
```

nashif commented:
@keith-zephyr  @sambhurst  is this something we are targeting for Zephyr 3.0? see https://github.com/zephyrproject-rtos/zephyr/projects/9

keith-zephyr commented:
> @keith-zephyr @sambhurst is this something we are targeting for Zephyr 3.0? see https://github.com/zephyrproject-rtos/zephyr/projects/9

I think so. This state machine framework will be used by USB-PD support, but the state machine is independent of USB-PD.  The full USB-PD stack is less likely to make it into Zephyr 3.0, but 3.1 could be a reasonable goal.

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -0,0 +1,317 @@
+	if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
+		internal->new_state = false;
+
+		if (smf_execute_ancestor_entry_actions(ctx, init_state)) {
```
Is this intended? You're executing the parent entry actions during set initial state.

sambhurst replied:
Yes this is intended and it's the same behavior as the ChromeOS SMF.

sambhurst commented:
@alexanderwachter thank you for your suggestions and at your earliest convenience, could you take another look at this PR?

alexanderwachter commented:
Is there a reason that the entry functions are called from the farthest parent to the nearest, but the run and exit in the opposite order? It seems that there is code duplication, and I tried to remove it. Then I stumbled over the issue above.

sambhurst commented:
In hierarchical state machines, entry actions are always executed starting with the outermost state and the run / exit actions are always executed in exact reverse order.

sambhurst commented:
@alexanderwachter Your suggestions greatly improved the framework, thank you. I've incorporated them into the code and added more tests. I also verified that the examples in the guide still work.

alexanderwachter commented:
LGTM now. Happy to help! I'm really looking forward to USB-PD support in Zephyr! Thanks for your work.

alexanderwachter commented:
@sambhurst I think your commit history has some issues. Please squash the commits

alexanderwachter commented:
👍 all good now!
